redhat-appstudio/service-provider-integration-operator
# second k8s client for searviceaccount requests

### What does this PR do?
Add second kubernetes client in oauth service that is used for Service Account requests (reading the secrets). Setting `API_SERVER` does not affect this new client.


### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

https://issues.redhat.com/browse/SVPI-291

### How to test this PR?


```
quay.io/mvala/service-provider-integration-operator:svpi291-apiserver
quay.io/mvala/service-provider-integration-oauth:svpi291-apiserver
```

I've tested on RHPDS openshift. I couldn't make it work on minikube due to invalid certificates.

The goal of the test is to verify that user's request (`SelfSubjectAccessReview`) goes to apiserver set by `API_SERVER` configuration, and other request (reading Secrets) goes directly to k8s apiserver. For that I've created tiny passthrough proxy that just logs requests https://gist.github.com/sparkoo/23bb8c5a5a58470b42c1c921e9e1abbf

1. login as cluster admin and deploy with images ^^^. No need to setup service providers, just keep it default.
2. deploy the proxy https://gist.github.com/sparkoo/23bb8c5a5a58470b42c1c921e9e1abbf#file-deployment-yaml you need to change `-upstream` arg to your cluster api url.
3. set `API_SERVER` in `spi-oauth-service-environment-config` configmap to host of testproxy Route, it will be something like `https://testproxy-spi-system.apps.cluster-4fxm9.4fxm9.sandbox1216.opentlc.com`. Useful to enable debug logs as well `ZAPDEVEL: "true"`. Restart oauth pod
4. login as user, create some openshift project
5. create SPIAccessTokenBinding `kubectl apply -f samples/spiaccesstokenbinding.yaml`
6. go through OAuth flow (login + go to OAuth URL). OAuth will fail because of configuration, but it does not matter here. Remember `clientId` from the URL.
7. you must see only single request `SelfSubjectAccessReview` in testproxy pod logs and you must see in oauth-service logs something like:
```
2023-01-31T08:16:17.343Z    DEBUG    config/userconfig.go:53    looking for sp configuration secrets    {"spHost": "github.com", "namespace": "user1", "spname": "GitHub", "sphost": "github.com"}                                                                                                                                                                           
2023-01-31T08:16:17.347Z    DEBUG    config/userconfig.go:71    found no secrets with sp configuration    {"spHost": "github.com"}
```
8. create service provider config secret in user's namespace `kubectl apply -f samples/sp-secret.yaml`
9. go to OAuth URL again, you must again see only single request in testproxy log, you must see different clientId in the URL, you must see in oauth-service logs something like
```
2023-01-31T08:17:02.136Z    DEBUG    config/userconfig.go:53    looking for sp configuration secrets    {"spHost": "github.com", "namespace": "user1", "spname": "GitHub", "sphost": "github.com"}                                                                                                                                                                           
2023-01-31T08:17:02.144Z    DEBUG    config/userconfig.go:75    found secrets with sp configuration    {"spHost": "github.com", "count": 1}
```
